### PR TITLE
SPI for Loadbalancer

### PR DESCRIPTION
**Pull request by tysonnorris**
To allow deployments to customize the Loadbalancer impls. Currently same Loadbalancer is used for all traffic, but future enhancements could use multiple load balancers for different traffic patterns.

Default is whisk.core.loadBalancer.LoadBalancerServiceProvider, which wraps the existing LoadBalancer impl  (Kafka + InvokerPool) behind the SPI for instantiation.

To decouple up the SPI interface for LoadBalancer from implementation details, no ArtifactStore stores are passed as parameters into LoadBalancerService anymore. CouchDbStoreProvider caches stores by name, so they can be fetched via SPI in multiple places and resolve to the same store instance.  Now LoadBalancerService can do `WhiskEntityStore.datastore(config)` to acquire the entityStore.

### CODE REVIEW

**Comment by tysonnorris** [comment]
A variety of tests are generating this exception ONLY in travis: `java.lang.IllegalStateException: Stream is terminated. SourceQueue is detached`

File: common/scala/src/main/scala/whisk/core/database/CouchDbStoreProvider.scala [hunk 1]
```diff
@@ -36,10 +38,17 @@ class CouchDbStoreProvider extends ArtifactStoreProvider {
         require(config.dbProvider == "Cloudant" || config.dbProvider == "CouchDB", "Unsupported db.provider: " + config.dbProvider)
         assume(Set(config.dbProtocol, config.dbHost, config.dbPort, config.dbUsername, config.dbPassword, name(config)).forall(_.nonEmpty), "At least one expected property is missing")
 
-        new CouchDbRestStore[D](config.dbProtocol, config.dbHost, config.dbPort.toInt, config.dbUsername, config.dbPassword, name(config))
+        val storeName = name(config)
+        stores.get(storeName) match {
```
Comment (markusthoemmes) [inline common/scala/src/main/scala/whisk/core/database/CouchDbStoreProvider.scala]: There's `getOrElseUpdate` on scala's maps. You can use it like this:

```scala
storeName.getOrElseUpdate(storeName, new CouchDbRestStore[D](config.dbProtocol, config.dbHost, config.dbPort.toInt, config.dbUsername, config.dbPassword, storeName))
```

File: core/controller/src/main/scala/whisk/core/loadBalancer/LoadBalancerService.scala [hunk 2]
```diff
@@ -77,16 +82,32 @@ trait LoadBalancer {
      */
     def publish(action: ExecutableWhiskAction, msg: ActivationMessage)(implicit transid: TransactionId): Future[Future[Either[ActivationId, WhiskActivation]]]
 
+    /**
+     * Return a message indicating the health of the containers and/or container pool in general
+     * @return a Future[String] representing the heal response that will be sent to the client
+     */
+    def getHealthResponse: Future[String]
```
Comment (markusthoemmes) [inline core/controller/src/main/scala/whisk/core/loadBalancer/LoadBalancerService.scala]: This should not be a `String` but rather a `JsObject` . It needs to contain the health of each invoker specifically.

File: common/scala/src/main/scala/whisk/core/database/CouchDbStoreProvider.scala (hunk 1 above)
Comment (tysonnorris) [inline common/scala/src/main/scala/whisk/core/database/CouchDbStoreProvider.scala]: great thanks!

File: core/controller/src/main/scala/whisk/core/loadBalancer/LoadBalancerService.scala (hunk 2 above)
Comment (tysonnorris) [inline core/controller/src/main/scala/whisk/core/loadBalancer/LoadBalancerService.scala]: done

**Comment by tysonnorris** [comment]
@markusthoemmes FYI - the test failures were due to some cases of ArtifactStore not being shutdown properly; fixed now

File: core/controller/src/main/scala/whisk/core/loadBalancer/LoadBalancerService.scala [hunk 3]
```diff
@@ -31,17 +31,19 @@ import akka.actor.ActorSystem
...
+import spray.json.DefaultJsonProtocol._
+import spray.json._
 import whisk.common.Logging
 import whisk.common.LoggingMarkers
 import whisk.common.TransactionId
 import whisk.core.WhiskConfig
 import whisk.core.WhiskConfig._
 import whisk.core.connector.MessagingProvider
-import whisk.core.connector.{ ActivationMessage, CompletionMessage }
+import whisk.core.connector.{ActivationMessage, CompletionMessage}
```
Comment (markusthoemmes) [inline core/controller/src/main/scala/whisk/core/loadBalancer/LoadBalancerService.scala]: Can you try to adjust this formatting?

File: common/scala/src/main/scala/whisk/core/database/CouchDbRestStore.scala [hunk 4]
```diff
@@ -52,7 +52,8 @@ class CouchDbRestStore[DocumentAbstraction <: DocumentSerializer](
     dbPort: Int,
     dbUsername: String,
     dbPassword: String,
-    dbName: String)(implicit system: ActorSystem, val logging: Logging, jsonFormat: RootJsonFormat[DocumentAbstraction])
+    dbName: String,
+    cleanup:(String)=>Unit)(implicit system: ActorSystem, val logging: Logging, jsonFormat: RootJsonFormat[DocumentAbstraction])
```
Comment (markusthoemmes) [inline common/scala/src/main/scala/whisk/core/database/CouchDbRestStore.scala]: Why do you need to pass the `dbName` in? When you open the client you automatically have it at hand.

`shutdown: => Unit` should suffice (note the whitespace please)

File: common/scala/src/main/scala/whisk/core/database/CouchDbStoreProvider.scala [hunk 5]
```diff
@@ -18,16 +18,18 @@
...
 import whisk.core.WhiskConfig
 import whisk.spi.Dependencies
-import whisk.spi.SpiFactory
+import whisk.spi.SingletonSpiFactory
 
 /**
  * A CouchDB implementation of ArtifactStoreProvider
  */
 class CouchDbStoreProvider extends ArtifactStoreProvider {
+    protected[database] val stores =  mutable.Map [String, CouchDbRestStore[_]]()
```
Comment (markusthoemmes) [inline common/scala/src/main/scala/whisk/core/database/CouchDbStoreProvider.scala]: Whitespace:

```scala
protected[database] val stores =  mutable.Map[String, CouchDbRestStore[_]]()
```

Could this be private?

File: core/controller/src/main/scala/whisk/core/loadBalancer/LoadBalancerService.scala [hunk 6]
```diff
@@ -275,6 +296,14 @@ class LoadBalancerService(
     private def generateHash(namespace: EntityName, action: ExecutableWhiskAction): Int = {
         (namespace.asString.hashCode() ^ action.fullyQualifiedName(false).asString.hashCode()).abs
     }
+
+    /** Returns a Map of invoker instance -> invoker state */
+    override def getHealthResponse(): Future[JsObject] = {
+        val res = allInvokers.map(_.map {
+            case (instance, state) => s"invoker${instance.toInt}" -> state.asString
+        }.toMap.toJson.asJsObject)
+        res
```
Comment (markusthoemmes) [inline core/controller/src/main/scala/whisk/core/loadBalancer/LoadBalancerService.scala]: There is no need to store the intermediary result:

```scala
override def getHealthResponse(): Future[JsObject] = allInvokers.map(_.map {
        case (instance, state) => s"invoker${instance.toInt}" -> state.asString
    }.toMap.toJson.asJsObject)
```

Also please refrain from using java style getter names like `getHealthResponse`. I suggest `LoadBalancer.healthStatus()`.

File: core/controller/src/main/scala/whisk/core/loadBalancer/LoadBalancerService.scala [hunk 7]
```diff
@@ -77,16 +82,32 @@ trait LoadBalancer {
...
     config: WhiskConfig,
-    instance: InstanceId,
-    entityStore: EntityStore)(
+    instance: InstanceId)(
         implicit val actorSystem: ActorSystem,
         logging: Logging)
     extends LoadBalancer {
 
+    /** Used to manage an action for testing invoker health */
+    val entityStore =  WhiskEntityStore.datastore(config)
```
Comment (markusthoemmes) [inline core/controller/src/main/scala/whisk/core/loadBalancer/LoadBalancerService.scala]: What's the rationale for generating a new one (and deduplicate via the SPI) vs. passing it through the `Dependencies` object? (Not necessarily arguing against this, just curiosity)

File: common/scala/src/main/scala/whisk/core/database/CouchDbStoreProvider.scala (hunk 5 above)
Comment (tysonnorris) [inline common/scala/src/main/scala/whisk/core/database/CouchDbStoreProvider.scala]: yes

File: core/controller/src/main/scala/whisk/core/loadBalancer/LoadBalancerService.scala (hunk 7 above)
Comment (tysonnorris) [inline core/controller/src/main/scala/whisk/core/loadBalancer/LoadBalancerService.scala]: It is not generating a new one, but acquiring a reference to the singleton EntityStore. The reference from LoadBalancerServer -> EntityStore is an impl detail (used to create the health action) that doesn't have anything to do with the LoadBalancer interface. To pass as Dependencies, the `SpiLoader.get` call in Controller would need to "know" this impl detail.

**Comment by tysonnorris** [comment]
@markusthoemmes thinking about this a bit, it seems prudent to include changes here to allow routing particular "action kinds" to particular load balancers (and therefore also enabling multiple load balancers), as briefly discussed in the meeting regarding concurrency. Let me know if you have an opinion on including that in this PR.

File: common/scala/src/main/scala/whisk/core/database/CouchDbStoreProvider.scala [hunk 8]
```diff
@@ -36,10 +38,14 @@ class CouchDbStoreProvider extends ArtifactStoreProvider {
         require(config.dbProvider == "Cloudant" || config.dbProvider == "CouchDB", "Unsupported db.provider: " + config.dbProvider)
         assume(Set(config.dbProtocol, config.dbHost, config.dbPort, config.dbUsername, config.dbPassword, name(config)).forall(_.nonEmpty), "At least one expected property is missing")
 
-        new CouchDbRestStore[D](config.dbProtocol, config.dbHost, config.dbPort.toInt, config.dbUsername, config.dbPassword, name(config))
+        val storeName = name(config)
+        stores.getOrElseUpdate(storeName, new CouchDbRestStore[D](config.dbProtocol, config.dbHost, config.dbPort.toInt, config.dbUsername, config.dbPassword, storeName,
+            ()=>stores.remove(storeName)))
+                .asInstanceOf[CouchDbRestStore[D]]
```
Comment (markusthoemmes) [inline common/scala/src/main/scala/whisk/core/database/CouchDbStoreProvider.scala]: Formatting this in a bit more context would make sense imho:

```scala
stores.getOrElseUpdate(storeName, 
    new CouchDbRestStore[D](config.dbProtocol, config.dbHost, config.dbPort.toInt, config.dbUsername, config.dbPassword, storeName, () => stores.remove(storeName))
).asInstanceOf[CouchDbRestStore[D]]
```

**Comment by markusthoemmes** [comment]
@tysonnorris I agree. A router for N-loadbalancer can be a subsequent step. Wouldn't include it in this PR.

File: common/scala/src/main/scala/whisk/core/database/CouchDbStoreProvider.scala [hunk 9]
```diff
@@ -18,16 +18,18 @@
...
 import whisk.core.WhiskConfig
 import whisk.spi.Dependencies
-import whisk.spi.SpiFactory
+import whisk.spi.SingletonSpiFactory
 
 /**
  * A CouchDB implementation of ArtifactStoreProvider
  */
 class CouchDbStoreProvider extends ArtifactStoreProvider {
+    private val stores =  mutable.Map [String, CouchDbRestStore[_]]()
```
Comment (markusthoemmes) [inline common/scala/src/main/scala/whisk/core/database/CouchDbStoreProvider.scala]: This should be a `TrieMap` to make it threadsafe.

**Comment by markusthoemmes** [comment]
This needs a rebase to reflect the latest changes to SPILoader.

File: tests/src/test/scala/whisk/core/database/test/CouchDBStoreProviderTests.scala [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+import whisk.core.WhiskConfig._
+import whisk.core.database.ArtifactStoreProvider
+import whisk.core.entity.WhiskActivation
+import whisk.spi.SpiLoader
+
+@RunWith(classOf[JUnitRunner])
+class CouchDBStoreProviderTests extends FlatSpec
+        with Matchers
+        with WskActorSystem
+        with StreamLogging {
```
Comment (markusthoemmes) [inline tests/src/test/scala/whisk/core/database/test/CouchDBStoreProviderTests.scala]: `StreamLogging` shouldn't be needed here.

File: tests/src/test/scala/whisk/core/database/test/CouchDBStoreProviderTests.scala [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+    override def afterAll() {
+        println("Shutting down store connections")
+        store1FirstLoad.shutdown()
+        //do not need to shutdown the second loaded store, but we'll do it anyways since clients may do that also
+        store1SecondLoad.shutdown()
+        super.afterAll()
+    }
+    it should "load the same store from config for a specific artifact type" in {
+        store1FirstLoad shouldBe store1SecondLoad
+    }
```
Comment (markusthoemmes) [inline tests/src/test/scala/whisk/core/database/test/CouchDBStoreProviderTests.scala]: Given the more general implementation you had in another PR (as discussed in Slack), should this be tested in a more general way? Also: Should we add a test which asserts that two stores with different names are different instances?

File: common/scala/src/main/scala/whisk/core/database/CouchDbStoreProvider.scala (hunk 9 above)
Comment (tysonnorris) [reply common/scala/src/main/scala/whisk/core/database/CouchDbStoreProvider.scala]: replaced this with SpiInstanceCaching (with TrieMap)

File: tests/src/test/scala/whisk/core/database/test/CouchDBStoreProviderTests.scala (hunk 10 above)
Comment (tysonnorris) [reply tests/src/test/scala/whisk/core/database/test/CouchDBStoreProviderTests.scala]: WhiskConfig requires `(implicit val logging: Logging)` - looks like StreamLogging is what most tests do. Please advise.

File: tests/src/test/scala/whisk/core/database/test/CouchDBStoreProviderTests.scala (hunk 11 above)
Comment (tysonnorris) [reply tests/src/test/scala/whisk/core/database/test/CouchDBStoreProviderTests.scala]: Sure; merge the SpiInstanceCaching to this PR; added test for different names -> different instances; will add general test for SpiInstanceCaching

**Comment by rabbah** [comment]
@tysonnorris also need labels here 🙏 .

**Comment by tysonnorris** [comment]
closing this per discussion with @markusthoemmes; will revise and submit a new PR in the near future

**Pull request closed without merging**